src/coordtest/src/lib.rs

```diff
@@ -0,0 +1,335 @@
+            _verbose: std::env::var_os("COORDTEST_VERBOSE").is_some(),
+            timestamp,
+        };
+        Ok(coordtest)
+    }
+
+    async fn connect(&self) -> anyhow::Result<SessionClient> {
+        let conn_client = self.client.as_ref().unwrap().new_conn()?;
+        let session = Session::new(conn_client.conn_id(), "materialize".into());
+        let (session_client, _startup_response) = conn_client.startup(session).await?;
+        Ok(session_client)
+    }
+
+    fn rewrite_query(&self, query: &str) -> String {
+        let path = self.temp_dir.path().to_str().unwrap();
+        let query = query.replace("<TEMP>", path);
+        query
+    }
+
+    // This very odd signature is some of us figured out as a way to achieve
+    // this. Is there a better way to do this?
+    async fn with_sc<F, T>(&self, f: F) -> T
+    where
+        F: for<'a> FnOnce(&'a mut SessionClient) -> std::pin::Pin<Box<dyn Future<Output = T> + 'a>>,
+    {
+        let mut sc = self.connect().await.unwrap();
+        let r = f(&mut sc).await;
+        sc.terminate().await;
+        r
+    }
+
+    async fn drain_uppers(&mut self, exclude: &HashSet<GlobalId>) {
+        loop {
+            if let Some(Some(mut msg)) = self.inner_feedback_rx.recv().now_or_never() {
+                // Filter out requested ids.
+                if let WorkerFeedback::FrontierUppers(uppers) = &mut msg.message {
+                    uppers.retain(|(id, _data)| !exclude.contains(id));
+                }
+                self.inner_feedback_tx.send(msg).unwrap();
+            } else {
+                return;
+            }
+        }
+    }
+
+    async fn make_catalog(&self) -> Catalog {
+        let catalog = self
+            .with_sc(|sc| Box::pin(async move { sc.dump_catalog().await.unwrap() }))
+            .await;
+        let catalog: BTreeMap<String, coord::catalog::Database> =
+            serde_json::from_str(&catalog).unwrap();
+        Catalog(catalog)
+    }
+}
+
+#[derive(Debug)]
+struct Catalog(BTreeMap<String, coord::catalog::Database>);
+
+impl Catalog {
+    fn get(&self, name: &str) -> GlobalId {
+        let mut name = name.split('.');
+        let database = self.0.get(name.next().unwrap()).unwrap();
+        let schema = database.schemas.get(name.next().unwrap()).unwrap();
+        let item_name = name.next().unwrap();
+        let id: GlobalId = schema
+            .items
+            .get(item_name)
+            .unwrap_or_else(|| panic!("{} not found, have: {:?}", item_name, schema.items.keys()))
+            .clone();
+        assert!(name.next().is_none());
+        id
+    }
+}
+
+async fn sql(sc: &mut SessionClient, query: String) -> anyhow::Result<Vec<ExecuteResponse>> {
+    let stmts = sql_parser::parser::parse_statements(&query)?;
+    let num_stmts = stmts.len();
+    let mut results = vec![];
+    for stmt in stmts {
+        let _ = sc.start_transaction(Some(num_stmts)).await;
+        sc.declare("".into(), stmt, vec![]).await?;
```

Comment:
I don't get this line

Reply:
Added comments. It's binding the empty portal with no params.